### PR TITLE
Distroless python sdk

### PR DESCRIPTION
**Pull request by damondouglas**
Addresses #32817 adding Python SDK (non-ML) [distroless](https://github.com/GoogleContainerTools/distroless) variants. 

#### Latest vulnerability report

<details>

<summary>Expand for more details:</summary>

image

</details>

#### Testing

Prior to submitting this PR, ran:
```
./gradlew :sdks:python:test-suites:dataflow:py311:validatesDistrolessContainer -PpythonVersion=3.11
```

<details>
  <summary>Expand for more detail:</summary>

```
> Task :sdks:python:test-suites:dataflow:py311:validatesDistrolessContainer
============================= test session starts ==============================
platform linux -- Python 3.11.9, pytest-7.4.4, pluggy-1.5.0
rootdir: /usr/local/google/home/damondouglas/projects/distroless-python-sdk/sdks/python
configfile: pytest.ini
plugins: xdist-3.6.1, hypothesis-6.116.0, timeout-2.3.1, requests-mock-1.12.1
timeout: 600.0s
timeout method: signal
timeout func_only: False
collected 1 item

apache_beam/examples/wordcount_it_test.py .                              [100%]

======================== 1 passed in 306.46s (0:05:06) =========================
```

image

</details>
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/python/container/Dockerfile [hunk 1]
```diff
@@ -103,9 +103,24 @@ RUN if [ "$pull_licenses" = "true" ] ; then \
...
 RUN if [ "$pull_licenses" != "true" ] ; then \
       rm -rf /opt/apache/beam/third_party_licenses ; \
     fi
+
+ARG TARGETARCH
+FROM gcr.io/distroless/python3-debian12:latest-${TARGETARCH} as distroless
+ARG py_version
+COPY --from=base /usr/local/include /usr/local/include
+COPY --from=base /usr/local/bin /usr/local/bin
+COPY --from=base /usr/local/lib /usr/local/lib
```
Comment (damondouglas) [inline sdks/python/container/Dockerfile:118]: Add comments as to why these directories are needed.

File: sdks/python/test-suites/dataflow/common.gradle [hunk 2]
```diff
@@ -380,6 +380,50 @@ task validatesContainer() {
...
+  def pyversion = "${project.ext.pythonVersion.replace('.', '')}"
+  dependsOn 'initializeForDataflowJob'
+  def pyProject = project(":sdks:python:container:py${pyversion}")
+  pyProject.ext['container-build-target'] = 'distroless'
+  rootProject.ext['docker-repository-root'] = "us.gcr.io/apache-beam-testing/${System.getenv('USER')}"
+  rootProject.ext['docker-tag'] = System.currentTimeMillis()
+  rootProject.ext['push-containers'] = true
+  dependsOn ":sdks:python:container:py${pyversion}:docker"
+
+  // TODO(damondouglas): the following is duplicate from container/common.gradle; attempts to single source via ext failed
```
Comment (damondouglas) [inline sdks/python/test-suites/dataflow/common.gradle:393]: Create Beam issue and reference in TODO

File: sdks/python/test-suites/dataflow/build.gradle [hunk 3]
```diff
@@ -60,6 +60,12 @@ task validatesContainerTests {
   }
 }
 
+task validateDistrolessContainerTests {
+  getVersionsAsList('distroless_python_versions').each {
+    dependsOn.add(":sdks:python:test-suites:dataflow:py${getVersionSuffix(it)}:validatesDistrolessContainer")
+  }
+}
+
```
Comment (damondouglas) [inline sdks/python/test-suites/dataflow/build.gradle:68]: Run this gradle task and submit results in PR description

**Review by Abacn — COMMENTED** [review]
Thanks, this is a good start. Note that the current change does not enter new automatically ran workflows. Could consider add a "beam_PostCommit_Python_ValidatesContainer_Dataflow_Distroless.yml" GitHub Action workflow, or just create another job running distroless containers in the existing "beam_PostCommit_Python_ValidatesContainer_Dataflow.yml" workflow

**Comment by damondouglas** [comment]
> Thanks, this is a good start. Note that the current change does not enter new automatically ran workflows. Could consider add a "beam_PostCommit_Python_ValidatesContainer_Dataflow_Distroless.yml" GitHub Action workflow, or just create another job running distroless containers in the existing "beam_PostCommit_Python_ValidatesContainer_Dataflow.yml" workflow

Thank @Abacn , the GitHub Action workflow will be addressed in #33003. The goal of this PR is to make backward compatible changes to existing Python container-related Dockerfile and related gradle tasks. Could you let me know if adding the workflow is important to your review of this PR and I can do so when I return from vacation in a couple weeks?

File: sdks/python/test-suites/dataflow/common.gradle [hunk 4]
```diff
@@ -380,6 +380,51 @@ task validatesContainer() {
...
+ * See https://cwiki.apache.org/confluence/display/BEAM/Python+Tips#PythonTips-VirtualEnvironmentSetup
+ * for more information on setting up different Python versions.
+ */
+task validatesDistrolessContainer() {
+  def pyversion = "${project.ext.pythonVersion.replace('.', '')}"
+  def buildTarget = 'distroless'
+  def repository = "us.gcr.io/apache-beam-testing/${System.getenv('USER')}"
+  def tag = java.time.Instant.now().getEpochSecond()
+  def imageURL = "${repository}/beam_python${project.ext.pythonVersion}_sdk_${buildTarget}:${tag}"
+  project.rootProject.ext['docker-repository-root'] = repository
```
[the remark this answers is not in the document]
Comment (damondouglas) [reply sdks/python/test-suites/dataflow/common.gradle:397]: @Abacn , thank you for meeting with me a couple days back. I realized that having separate `py<version>-distroless` gradle projects could be avoided with the logic introduced in container/common.gradle by querying the `container-build-target` property. Thus, there is no `py<version>-distroless`. The description has a screenshot in the collapsed markdown of the Dataflow Job that succeeded with the `validatesDistrolessContainer` task. I don't have the means while currently on vacation to test again but I (mostly) remember that the tag is needed to be set on the rootProject level to work. `validatesDistrolessContainer` should not interfere with other tasks and these properties should only be set when the task executes, not when it is configured.

**Review by Abacn — APPROVED** [review]
thanks for the clarifications. This looks good to me. I understand this is currently an no-op to workflow as both test workflow and push docker image to dockerhub are pending. Defer to you for the timeline to merge

**Pull request merged**